test/gtest/arch_test.cpp

```diff
@@ -0,0 +1,348 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "gtest/gtest.h"
+#include "ir/ir.h"
+#include "helpers.h"
+#include "lib/log.h"
+
+#include "frontends/p4/typeMap.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/common/resolveReferences/resolveReferences.h"
+
+#include "p4/createBuiltins.h"
+#include "p4/typeChecking/typeChecker.h"
+
+using namespace P4;
+
+TEST(arch, packet_out) {
+    std::string program = R"(
+        // simplified core.p4
+        extern packet_out {
+            void emit<T> (in T hdr);
+        }
+        // simplified v1model.p4
+        control Deparser<H> (packet_out b, in H hdr);
+        package PSA<H> (Deparser<H> p);
+        // user program
+        struct ParsedHeaders {
+            bit<32> hdr;
+        }
+        control MyDeparser(packet_out b, in ParsedHeaders h){
+            apply {
+                b.emit(h.hdr);
+            }
+        }
+        PSA(MyDeparser()) main;
+    )";
+    const IR::P4Program* pgm = parse_string(program);
+
+    ReferenceMap refMap;
+    TypeMap      typeMap;
+    PassManager  passes({
+        new TypeChecking(&refMap, &typeMap)
+    });
+
+    pgm = pgm->apply(passes);
+    ASSERT_NE(nullptr, pgm);
+}
+
+// Potential bug
+TEST(arch, duplicatedDeclarationBug) {
+    std::string program = R"(
+        // simplified core.p4
+        extern packet_out {
+            void emit<T> (in T hdr);
+        }
+        // simplified v1model.p4
+        control Deparser<H> (packet_out b, in H hdr);
+        package PSA<H> (Deparser<H> p);
+        // user program
+        struct ParsedHeaders {
+            bit<32> hdr;
+        }
+        // BUG: duplicated name with 'Deparser'
+        control Deparser(packet_out b, in ParsedHeaders h){
```

Comment:
I believe that this is truly wasteful.
It is *much* simpler to just add a file containing the two declarations into testdata/p4_16_errors.

This will not only test that the compilation fails, but it will also check that the error message is the correct one. It also requires much less code: no c++, no "simplified" core.p4.